controllers/upgrade_controller.go

```diff
@@ -0,0 +1,231 @@
+)
+
+const OfedDriverLabel = "nvidia.com/ofed-driver"
+
+// UpgradeReconciler reconciles OFED Daemon Sets for upgrade
+type UpgradeReconciler struct {
+	client.Client
+	Log    logr.Logger
+	Scheme *runtime.Scheme
+}
+
+// ResourceType holds information about k8s resource type
+type ResourceType struct {
+	APIVersion string
+	Kind       string
+}
+
+//nolint
+// +kubebuilder:rbac:groups=mellanox.com,resources=*,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups="",resources=nodes,verbs=get;list;watch;update;patch
+// +kubebuilder:rbac:groups="",resources=pods,verbs=list
+// +kubebuilder:rbac:groups=apps,resources=deployments;daemonsets;replicasets;statefulsets,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=apps,resources=deployments/finalizers,verbs=update
+
+// Reconcile is part of the main kubernetes reconciliation loop which aims to
+// move the current state of the cluster closer to the desired state.
+func (r *UpgradeReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	reqLogger := r.Log.WithValues("upgrade", req.NamespacedName)
+	reqLogger.V(consts.LogLevelInfo).Info("Reconciling Upgrade")
+
+	nicClusterPolicy := &mellanoxv1alpha1.NicClusterPolicy{}
+	err := r.Get(context.TODO(), types.NamespacedName{Name: consts.NicClusterPolicyResourceName}, nicClusterPolicy)
+
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	upgradePolicy := nicClusterPolicy.Spec.OfedUpgradePolicy
+	if upgradePolicy == nil {
+		reqLogger.V(consts.LogLevelInfo).Info("OFED Upgrade Policy is not defined, skipping driver upgrade")
+		return ctrl.Result{}, nil
+	}
+
+	state, err := r.BuildState(ctx, upgradePolicy)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	reqLogger.V(consts.LogLevelDebug).Info("Built upgrade state", "state", state)
+
+	return ctrl.Result{}, nil
+}
+
+func (r *UpgradeReconciler) BuildState(
+	ctx context.Context, upgradePolicy *mellanoxv1alpha1.OfedUpgradePolicySpec) (*upgrade.ClusterUpgradeState, error) {
+	r.Log.V(consts.LogLevelInfo).Info("Building state", "Upgrade policy:", upgradePolicy)
+
+	upgradeState := upgrade.NewClusterUpgradeState()
+	upgradeState.Policy = upgradePolicy
+
+	daemonSets, err := r.getDriverDaemonSets(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	r.Log.V(consts.LogLevelDebug).Info("Got driver daemon sets", "length", len(daemonSets))
+
+	// Get list of driver pods
+	podList := &corev1.PodList{}
+
+	err = r.List(ctx, podList,
+		client.InNamespace(consts.NetworkOperatorResourceNamespace), client.MatchingLabels{OfedDriverLabel: ""})
+	if err != nil {
+		return nil, err
+	}
+
+	for i := range podList.Items {
+		pod := &podList.Items[i]
+		r.Log.V(consts.LogLevelDebug).Info("Pod", "pod", pod.Name, "owner", pod.OwnerReferences[0].Name)
+
+		if pod.OwnerReferences == nil || len(pod.OwnerReferences) < 1 {
+			r.Log.V(consts.LogLevelWarning).Info("OFED Driver Pod has no owner DaemonSet", "pod", pod)
```

Comment:
can this happen for ofed pod ?

Reply:
It shouldn't but I'd still check for this case to avoid potential access violation